src/crystal/system/win32/socket.cr

```diff
@@ -0,0 +1,425 @@
+      LibC.WSASend(fd, pointerof(wsabuf), 1, out bytes_sent, 0, overlapped, nil)
+    end
+
+    bytes.to_i32
+  end
+
+  private def system_send_to(bytes : Bytes, addr : ::Socket::Address)
+    wsabuf = wsa_buffer(bytes)
+    bytes_sent = overlapped_write(fd, "WSASendTo") do |overlapped|
+      LibC.WSASendTo(fd, pointerof(wsabuf), 1, out bytes_sent, 0, addr, addr.size, overlapped, nil)
+    end
+    raise ::Socket::Error.from_errno("Error sending datagram to #{addr}") if bytes_sent == -1
+
+    # to_i32 is fine because string/slice sizes are an Int32
+    bytes_sent.to_i32
+  end
+
+  private def system_receive(bytes)
+    sockaddr = Pointer(LibC::SOCKADDR_STORAGE).malloc.as(LibC::Sockaddr*)
+    # initialize sockaddr with the initialized family of the socket
+    copy = sockaddr.value
+    copy.sa_family = family
+    sockaddr.value = copy
+
+    addrlen = sizeof(LibC::SOCKADDR_STORAGE)
+
+    wsabuf = wsa_buffer(bytes)
+
+    flags = 0_u32
+    bytes_read = overlapped_read(fd, "WSARecvFrom") do |overlapped|
+      LibC.WSARecvFrom(fd, pointerof(wsabuf), 1, out bytes_received, pointerof(flags), sockaddr, pointerof(addrlen), overlapped, nil)
+    end
+
+    {bytes_read.to_i32, sockaddr, addrlen}
+  end
+
+  private def system_close_read
+    if LibC.shutdown(fd, LibC::SH_RECEIVE) != 0
+      raise ::Socket::Error.from_errno("shutdown read")
+    end
+  end
+
+  private def system_close_write
+    if LibC.shutdown(fd, LibC::SH_SEND) != 0
+      raise ::Socket::Error.from_errno("shutdown write")
+    end
+  end
+
+  private def system_reuse_port?
+    false
+  end
+
+  private def system_reuse_port=(val : Bool)
+    raise NotImplementedError.new("Socket#reuse_port=")
+  end
+
+  private def system_linger
+    v = LibC::Linger.new
+    ret = getsockopt LibC::SO_LINGER, v
+    ret.l_onoff == 0 ? nil : ret.l_linger
+  end
+
+  private def system_linger=(val)
+    v = LibC::Linger.new
+    case val
+    when Int
+      v.l_onoff = 1
+      v.l_linger = val
+    when nil
+      v.l_onoff = 0
+    end
+
+    setsockopt LibC::SO_LINGER, v
+    val
+  end
+
+  def system_getsockopt(handle, optname, optval, level = LibC::SOL_SOCKET)
+    optsize = sizeof(typeof(optval))
+    ret = LibC.getsockopt(handle, level, optname, pointerof(optval).as(UInt8*), pointerof(optsize))
+
+    if ret.zero?
+      yield optval
+    else
+      raise ::Socket::Error.from_wsa_error("getsockopt #{optname}")
+    end
+
+    ret
+  end
+
+  def system_setsockopt(handle, optname, optval, level = LibC::SOL_SOCKET)
+    optsize = sizeof(typeof(optval))
+
+    ret = LibC.setsockopt(handle, level, optname, pointerof(optval).as(UInt8*), optsize)
+    raise ::Socket::Error.from_wsa_error("setsockopt #{optname}") if ret == LibC::SOCKET_ERROR
+    ret
+  end
+
+  private def system_blocking?
+    mode = uninitialized UInt32
+    LibC.WSAIoctl(fd, LibC::FIONBIO, nil, 0, pointerof(mode), sizeof(UInt32), out bytes_returned, nil, nil)
```

Comment:
Check for [errors in the call to WSAIoctl](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsaioctl#return-value)

Reply:
Oh, thanks for pointing out. This is actually a major flaw because that call always errored. Apparently, you cant use `WSAIoctl` to query whether the socket is blocking or non-blocking.
https://stackoverflow.com/questions/3974193/how-to-find-out-if-a-socket-is-already-in-non-blocking-mode-in-windows

So it looks like the best option excluding any more complicated hacks to find out what's what, is to just keep track of blocking status in an instance variable. That's not completely sane because the blocking mode could be changed directly without going through `Socket#blocking=`. But, that's fairly off the track.

Reply:
Ah, I see. Seems like we have to figure out a way around that one way or the other.